### PR TITLE
More detailed SLA miss email

### PR DESCRIPTION
**Pull request by patrickleotardif**
@mistercrunch Basically I want to have an email that very simply states why the DAG didn't complete on time. (e.g. 'dim_users is not done yet') 

I'm not sure how to test this, so advice in that respect welcome!
Also, if this is too specific I guess there are many other ways I could build this.

(For context, I want this as a building block for a single UI to schedule a SQL query, this UI would ideally have an automatic SLA built in that would email the end user when their query doesn't finish on time as usual, essentially automating the process of them emailing a poor data engineer who simply says ' yeah dim_users didn't land on time')

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -272,13 +272,29 @@ def manage_slas(self, dag, session=None):
...
             .all()
         )
+
+        current_run_date = datetime.now() - dag.schedule_interval
+        blocking_tis = (
+            session
+            .query(TI)
+            .filter(TI.state != State.SUCCESS)
+            .filter(TI.execution_date == current_run_date)
+            .filter(TI.are_dependencies_met())
```
Comment (mistercrunch) [inline airflow/jobs.py]: I bet this line doesn't work. `query.filter` and `query.fliter_by` are ORM methods that are part of figuring out what SQL to generate, can't bring python logic into here...

You'll have to retrieve the task instances and then run `are_dependencies_met`

Comment (patrickleotardif) [inline airflow/jobs.py]: i.e. I remove that line from blocking_tis and then do

```
blocking_tis = [ti for ti in blocking_tis if ti.are_dependencies_met()]
```

?

Comment (patrickleotardif) [inline airflow/jobs.py]: do I need to do session.commit() or anything?

**Comment by patrickleotardif** [comment]
Any advice on testing this function manually @mistercrunch ?

File: airflow/jobs.py (hunk 1 above)
Comment (mistercrunch) [inline airflow/jobs.py]: Should be good. No need to commit until values are changed and `session.merge` is called

**Comment by mistercrunch** [comment]
I think the way I tested the SLA feature before was by setting an artificial SLA on a specific DAG that would be partially processed and running the scheduler only for that DAG (`airflow scheduler -d foo_dag`). You need to have email configured properly on that box to actually test it.

**Comment by patrickleotardif** [comment]
@mistercrunch I fixed your notes and tested it fully on prod by triggering the scheduler in the middle of backfills that I created for test dags.

File: airflow/jobs.py [hunk 2]
```diff
@@ -272,13 +272,33 @@ def manage_slas(self, dag, session=None):
             .filter(SlaMiss.dag_id == dag.dag_id)
             .all()
         )
+
+        current_run_date = (datetime.now() -
+                            dag.schedule_interval).isoformat()[:10]
```
Comment (mistercrunch) [inline airflow/jobs.py]: This seems arbitrary, wouldn't work with non-daily DAG, can we just work off of the list we're about to send an email about?

Comment (patrickleotardif) [inline airflow/jobs.py]: how about 

```
sla_dates = [sla.execution_date.isoformat()[:10] for sla in slas]

(...)
 .filter(TI.execution_date.in(sla_dates))
```

**Comment by patrickleotardif** [comment]
Added changes as discussed and sent some test emails on prod successfully (across multiple days too).

**Comment by patrickleotardif** [comment]
@mistercrunch I would merge but I can't :cry:

**Comment by mistercrunch** [comment]
Mergin'!

**Pull request merged**